<a href="https://colab.research.google.com/github/reshmi56/Hospital-/blob/main/Pandas_tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

To work with large amount of dataset

In [1]:
import pandas as pd
df=pd.read_csv('/content/heart_records_dataset.csv')
df

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                       582         0                 20   
1    55.0        0                      7861         0                 38   
2    65.0        0                       146         0                 20   
3    50.0        1                       111         0                 20   
4    65.0        1                       160         1                 20   
..    ...      ...                       ...       ...                ...   
294  62.0        0                        61         1                 38   
295  55.0        0                      1820         0                 38   
296  45.0        0                      2060         1                 60   
297  45.0        0                      2413         0                 38   
298  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1  265000.00               1.9           130    1   
1                      0  263358.03               1.1           136    1   
2                      0  162000.00               1.3           129    1   
3                      0  210000.00               1.9           137    1   
4                      0  327000.00               2.7           116    0   
..                   ...        ...               ...           ...  ...   
294                    1  155000.00               1.1           143    1   
295                    0  270000.00               1.2           139    0   
296                    0  742000.00               0.8           138    0   
297                    0  140000.00               1.4           140    1   
298                    0  395000.00               1.6           136    1   

     smoking  time  DEATH_EVENT  
0          0     4            1  
1          0     6            1  
2          1     7            1  
3          0     7            1  
4          0     8            1  
..       ...   ...          ...  
294        1   270            0  
295        0   271            0  
296        0   278            0  
297        1   280            0  
298        1   285            0  

[299 rows x 13 columns]

**Dataframe**
2 dimension array or table with rows and columns

**Different ways of creating dataframe**
csv,excel,dictionary,tuple

---------csv----------
import pandas as pd
df=pd.read_csv('')
df

--------------excel-----------
df=pd.read_excel('')
df

--------------dictionary-------
weather={
         'day'=[mon,tues]
         'color'=[red,blue]
}
df=pd.DataFrame(weather)
df

-------tuple---------
weather=[
  ('xx',23)
]
df=pd.DataFrame(data=weather,columns=['name','age'])
df


**Handling missing values**

fillna--->ffill,bfill

dropna--thresh

interpolate---->linear,time based

**Handle missing date**

dt=pd.date_range()

idx=pd.DatetimeIndex(dt)

df.reindex(dt)

**Handle using replace**

new=pd.replace(to_replace=[-999],value=0)

new

new=pd.replace({
  -999:np.nan
  'no event':'sunday'})

  new=pd.replace({
    'temp':-99
  }np.nan)

**GroupBy**
Split ,Apply ,Combine

g=df.groupby("city")

for city,data in g:
  print(city)
  print(data)

g.max()


**concat()**

df=pd.concat([us,ch],ignore_index=True)

df=pd.concat([us,ch],keys=['usa','china'])

df.loc(us)

**merge()**

new=pd.merge(df1,df2,how='outer',suffixes=_x,_y,indicator=True)(gives which table)

outer,inner,left,right

**pivot()**
pivot allows to transform or reshape data

**pivottable**
allows to summarize and aggregate data inside dataframe

DataFrame.pivot(index=None, columns=None, values=None)

DataFrame.pivot_table(values=None, index=None, columns=None, aggfunc='mean', fill_value=None)


**melt()**
converts wide format dataframe into long format dataframe

import pandas as pd

# Sample Data
data = {'Category': ['A', 'B'], 'Type_X': [10, 30], 'Type_Y': [20, 40]}

df = pd.DataFrame(data)

# Melt
melted = df.melt(id_vars='Category', var_name='Type', value_name='Value')


**Stack**
converts columns to rows(long data format)

for data analysis

**Unstack**
converts rows to columns(wide data format)

for data visualisation

new=pd.DataFrame(df,index=['1','2'])

stacked=new.stack()

unstacked=stacked.unstack()




**CROSSTAB**

summarizing the relationship between different categories and for counting occurrences.

pandas.crosstab(index, columns, values=None, aggfunc='count', margins=False, margins_name='All', dropna=True)


margins (optional): If True, adds a summary row and column.
margins_name (optional): Name for the margins (default is 'All').

import pandas as pd

# Sample Data
data = {'Gender': ['Male', 'Female', 'Female', 'Male', 'Female', 'Male'],
        'Department': ['HR', 'Engineering', 'HR', 'Engineering', 'Engineering', 'HR']}
df = pd.DataFrame(data)

# Cross-tabulate Gender vs Department
ct = pd.crosstab(df['Gender'], df['Department'])
print(ct)

Department  Engineering  HR

Gender
                     
Female              2   1

Male                1   2


# Adding margins to include row and column totals
ct_with_margins = pd.crosstab(df['Gender'], df['Department'], margins=True)
print(ct_with_margins)

Department  Engineering  HR  All

Gender                          

Female              2   1    3

Male                1   2    3

All                 3   3    6


**Time Series Analysis**

set of datapoints indexed in time order

**Datetime Index**

**pd.to_datetime(),set_index()**

data={'date':[''],'value':['']}

df=pd.DataFrame(data)

df['date']=pd.to_datetime(df['date'])

df.set_index('date',inplace=True)

**Resampling**
converts to weekly monthly

weekly=df.resample('W').mean()

'D': Daily
'W': Weekly
'M': Monthly
'Q': Quarterly
'A': Annually

**Shifting**:

Shifts data by a specified time frequency, useful for comparisons between time periods.

df_shifted = df.shift(1)

print(df_shifted)

         Value
Date  

2023-01-01    NaN

2023-01-02    100

2023-01-03    200


**Rolling Window:**

Calculates statistics (e.g., mean, sum) over a rolling window of a specified size.

rolling_mean = df['Value'].rolling(window=3).mean()


Date
2023-01-01      NaN

2023-01-02      NaN

2023-01-03    150.0

Name: Value, dtype: float64



**Decomposition**

Trend: The general direction the data is moving (e.g., increasing, decreasing).

Seasonality: Regular patterns repeating over time (e.g., higher sales in December).

Residuals (Noise): The random fluctuations or irregularities.

import pandas as pd
from statsmodels.tsa.seasonal import seasonal_decompose

# Example Time Series Data
data = [100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 200]

dates = pd.date_range('2023-01-01', periods=len(data), freq='D')

ts = pd.Series(data, index=dates)

# Decompose the time series
decomposition = **seasonal_decompose**(ts, model='additive', period=3)

decomposition.plot()


**Time Series Forecasting:**

Autoregressive (AR), Moving Average (MA), and ARMA/ARIMA
models are used for forecasting future values based on past observations.

Forecasting can be done using libraries such as statsmodels or prophet.

from statsmodels.tsa.arima.model import ARIMA

# Example time series data
ts = pd.Series([100, 120, 130, 150, 170, 200, 220],
               index=pd.date_range('2023-01-01', periods=7, freq='D'))

# Fit ARIMA model
model = ARIMA(ts, order=(1, 1, 1))

model_fit = model.fit()

# Forecast next 3 values
forecast = model_fit.forecast(steps=3)

print(forecast)


**Series**: A one-dimensional labeled array capable of holding any data type

import pandas as pd

s = pd.Series([1, 2, 3, 4])

print(s)


**loc**: Access rows and columns by labels.

df.loc[0, 'Name']  # Access first row, 'Name' column

**iloc**: Access rows and columns by positions.

df.iloc[0, 1]  # Access first row, second column

at: Faster access for scalar values by label.

df.at[0, 'Name']

iat: Faster access for scalar values by position.

df.iat[0, 1]

Filtering Data:

Boolean Indexing:

df[df['Age'] > 25]

Querying:

df.query("Age > 25")

**apply**(): Apply a function along an axis (row/column).

df['Age'].apply(lambda x: x * 2)

**map**(): Apply a function element-wise to a Series.

df['Age'].map(lambda x: x * 2)

**Vectorized Operations:**

df['Age'] * 2

**Sorting**:
**By values:**

df.sort_values('Age', ascending=False)

**By index:**

df.sort_index()

**Large Datasets:**

Process in chunks:

for chunk in pd.read_csv('file.csv', chunksize=1000):
    process(chunk)

**Memory Usage:**

Check:

df.memory_usage()

df = pd.read_csv('file.csv')

df.to_csv('file.csv')
